client/src/wayland_obj/manager.rs

```diff
@@ -0,0 +1,168 @@
+//     })
+// }
+
+pub fn create_buffer(size: Size) -> Result<Buffer, ()> {
+    WAYLAND.with(|wayland| {
+        let wayland = wayland.borrow();
+        let wayland = wayland.as_ref().expect("Wayland not initialized");
+
+        wayland.shm_manager.create_buffer(size)
+    })
+}
+
+pub fn create_layer_surface() -> Result<LayerSurface, ()> {
+    WAYLAND.with(|wayland| {
+        use layer_shell::Layer;
+        let wayland = wayland.borrow();
+        let wayland = wayland.as_ref().expect("Wayland not initialized");
+
+        let wl_surface = wayland.compositor_manager.create_surface()?;
+
+        wayland
+            .shell_manager
+            .create_layer_surface(wl_surface, None, Layer::Top)
+    })
+}
+
+pub fn global_callback(event: GlobalEvent, registry: WlRegistry) {
+    WAYLAND.with(|wayland| {
+        let wayland = &mut wayland.borrow_mut();
+        let wayland = wayland.as_mut().expect("Wayland not initialized");
+        match event {
+            GlobalEvent::New {
+                id,
+                interface,
+                version
+            } => wayland.new_global(registry, id, &interface, version),
+            _ => unimplemented!("GlobalEvent")
+        }
+    })
+}
+
+pub fn init_wayland() -> (Display, EventQueue) {
+    let (display, event_queue) = Display::connect_to_env().unwrap_or_else(|err| {
+        use std::{env, process::exit};
+        use wayland_client::ConnectError::*;
+        match err {
+            NoWaylandLib => error!("Could not find Wayland library, is it installed and in PATH?"),
+            NoCompositorListening => {
+                error!("Could not connect to Wayland server. Is it running?");
+                error!(
+                    "WAYLAND_DISPLAY={}",
+                    env::var("WAYLAND_DISPLAY").unwrap_or_default()
+                );
+            },
+            InvalidName => error!("Invalid socket name provided in WAYLAND_SOCKET"),
+            XdgRuntimeDirNotSet => error!("XDG_RUNTIME_DIR must be set"),
+            InvalidFd => error!("Invalid socket provided in WAYLAND_SOCKET")
+        }
+        exit(1);
+    });
+
+    WAYLAND.with(|wayland| {
+        let wayland = &mut wayland.borrow_mut();
+        let way_man = WaylandManager::new(crate::lua::WaylandHandler {});
+        (*wayland).replace(way_man);
+    });
+
+    (display, event_queue)
+}
+
+impl WaylandManager {
+    pub fn new<T>(handler: T) -> Self
+    where
+        T: OutputEventHandler + 'static
+    {
+        let handler = Rc::new(handler);
+        WaylandManager {
+            compositor_manager: WlCompositorManager::default(),
+            output_manager: WlOutputManager::new(handler),
+            seat_manager: WlSeatManager::new(),
+            shell_manager: LayerShellManager::new(),
+            shm_manager: WlShmManager::new()
+        }
+    }
+
+    fn new_global(&mut self, registry: WlRegistry, id: u32, interface: &str, version: u32) {
+        match interface {
```

Comment:
Originally I wanted a function like this:
```
fn get_handler(interface: String) -> (min_version: u32, implementor: &mut GlobalImplementor<Interface>) {
  match interface {
    WlCompositor::NAME => (WL_COMPOSITOR_VERSION, &mut self.compositor_manager),
    ....
  }
}
```
That could spare duplicate code when `GlobalEvent::Removed` gets implemented..
I have the impression that rust does not allow this, or did I miss something?

Reply:
This would return a pointer to a stack allocated object (since it was constructed in the body of the function and its drop will be called at the end of the stack frame). In particular this is a trait object (I believe some version of clippy should complain that this doesn't use `dyn`, which would make this more obvious...I should review how the CI is  set up). In order to return a trait object you need to box it explicitly, e.g return `Box<GlobalImplementor<Interface>>`. You could also use `impl GlobalImplementor<Interface>` IIRC instead.

Reply:
I believe the problem was that `Interface` is also a trait.

Reply:
Hm, that shouldn't matter. It should still be possible to box the outer trait.